internals/overlord/servstate/handlers.go

```diff
@@ -223,6 +223,9 @@ func (m *ServiceManager) doStop(task *state.Task, tomb *tomb.Tomb) error {
 	if service == nil {
 		return nil
 	}
+	// Close the ringbuffer to signal to consumers (e.g. log forwarding) that
+	// there are no more logs coming.
+	defer service.logs.Close()
```

Comment:
I believe this is the critical piece of plumbing that is required to make the pullers exit as planned during Pebble shutdown? This hook is the stop handler, so it will only be used by external stop requests, including the tear-down function of the daemon. However, that function will only stop services in ```running``` and ```backoff``` state. So I am wondering what about all the other services that may have been started in the past (have valid ring buffers), but which are not running right now? (e.g. crashed service with restart action = ignore). Any puller associated with such as service will cause the exit to delay by the puller timeout delay. Not immediately sure the best way to solve this, could the termination happens in the ```overlord.exited()``` hook, but not sure if can be reused after closing it, because ```exited()``` if often only a temp transition?

Reply:
Yeah, also if you stop a service which is in `starting` state (e.g. interrupting Pebble shortly after it starts), it doesn't work. Not sure where is the best place for this code.

Reply:
Spoke to @benhoyt about this, we agreed it makes the most sense to have the ring buffer lifecycle tied to the service data (i.e. when we no longer need the service data, we no longer need the ring buffer either). So we've put the close in `ServiceManager.removeService`, and are ensuring we call `removeService` in the `ServiceManager.Stop` code.